docs/source/app-dev/bindings-x-lang/index.rst

```diff
@@ -0,0 +1,97 @@
+.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+Creating your own bindings
+##########################
+
+This page gets you started with creating custom bindings for the Digital Asset distributed ledger.
+
+Introduction
+============
+
+Digital Asset currently provides bindings for the following programming languages:
+
+- :doc:`Java </app-dev/bindings-java/index>`
+
+- :doc:`Scala </app-dev/bindings-scala/index>`
+
+- :doc:`JavaScript (Node.js) </app-dev/bindings-js>`
+
+However, you can create bindings for any programming language supported by `gRPC <https://grpc.io/docs/>`_.
+
+What do we mean by "bindings"? Bindings for a language consist of two main components:
+
+- Ledger API
+    Client "stubs" for the programming language, -- the remote API that allows sending ledger commands and receiving ledger transactions. You have to generate **Ledger API** from `the gRPC protobuf definitions in the daml repository on GitHub <https://github.com/digital-asset/daml/tree/master/ledger-api/grpc-definitions>`_. **Ledger API** is documented on this page: :doc:`/app-dev/grpc/index`. The `gRPC <https://grpc.io/docs/>`_ tutorial explains how to generate client "stubs".
+
+- Codegen
+     A code generator is a program that generates classes representing DAML contract templates in the language. These classes incorporate all boilerplate code for constructing: :ref:`com.digitalasset.ledger.api.v1.CreateCommand` and :ref:`com.digitalasset.ledger.api.v1.ExerciseCommand` corresponding for each DAML contract template.
+
+Technically codegen is optional. You can construct the commands manually from the auto-generated **Ledger API** classes. However it is very tedious and error-prone. If you are creating ah hoc bindings for a project with a few contract templates, writing a proper codegen may be an overkill. On the other hand if you have hundreds of contract templates in your project or planning to build language bindings that you will share across multiple projects, we recommend including codegen in your bindings. It will save you time in the long run.
```

Comment:
It's a little more difficult to talk about, but there's really a broad category of metaprogramming features that can solve this problem just as well or even better than codegen; they're language-specific, but often much easier to maintain (i.e. no need to add a build step). Some examples are

1. F# [type providers](https://docs.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/creating-a-type-provider#a-type-provider-that-is-backed-by-local-data),
2. Template Haskell,
3. Scala macro annotations (not future-proof enough to use when implementing the last Scala codegen),

or pretty much any sort of metaprogramming mechanism that works opposite to the usual unsafe "reflection", working to _produce_ types instead of _inspect_ them, at _compile_ time instead of runtime.

I don't know how much we want to say about that, though.

Reply:
@S11001001 are you suggesting we remove this entire section of just leave it like:
> Technically codegen is optional. You can construct the commands manually from the auto-generated **Ledger API** classes. However it is very tedious and error-prone

Reply:
but I think we need to give readers a hint that you can solve the same problem with either codegen, macros or metaprogramming if you don't want to write your commands manually.

Reply:
@S11001001, how about this explantion for the `codgen`. A code generator is a program or any metaprogramming technic that generates ledger commands  at compile-time.

Reply:
> are you suggesting we remove this entire section

Not at all, you're right ⬇️ 

> we need to give readers a hint that you can solve the same problem with either codegen, macros or metaprogramming if you don't want to write your commands manually

Because nowhere else have we explained why we bother writing codegens, it's just something _we_ know.

Also, notwithstanding a rewrite to talk about other metaprogramming forms:

```suggestion
Technically codegen is optional. You can construct the commands manually from the auto-generated **Ledger API** classes. However, it is very tedious and error-prone. If you are creating *ad hoc* bindings for a project with a few contract templates, writing a proper codegen may be overkill. On the other hand, if you have hundreds of contract templates in your project or are planning to build language bindings that you will share across multiple projects, we recommend including a codegen in your bindings. It will save you and your users time in the long run.
```

Reply:
> how about this explantion for the `codgen`. A code generator is a program or any metaprogramming technic that generates ledger commands at compile-time.

No, that would be very misleading. I mentioned the other metaprogramming techniques because they _aren't_ codegens.

Reply:
added codegen alternatives